File: 45-jump-game-ii/45-jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        if (nums.size() == 1){
            return 0;
        }
        int maxReach = nums[0];
        int steps = 1;
        int i = 0;
        while (i < nums.size()){
            if (maxReach >= nums.size() - 1){
                return steps;
            }
            int temp = 0;
            while (i < nums.size() && i <= maxReach){
                temp = max(temp, i + nums[i]);
                i++;
            }
            maxReach = temp;
            steps++;
        }
        return steps;
    }
};


// # Single loop O(n) Solution
// if len(nums) == 1:
//     return 0
// maxReach, steps = nums[0], 1
// i = 0
// while i < len(nums):
//     if maxReach >= len(nums) - 1:
//         return steps
//     temp = 0
//     while i < len(nums) and i <= maxReach:
//         temp = max(temp, i + nums[i])
//         i += 1
//     maxReach = temp
//     steps += 1
// return steps

File: 45-jump-game-ii/45-jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        @cache
        def chooseIndex(i):
            if i >= len(nums) - 1:
                return 0
            ans = float("inf")
            for j in range(i + 1, i + nums[i] + 1):
                ans = min(ans, chooseIndex(j) + 1)
            return ans
        return chooseIndex(0)